File: Windows/src/IwashiScope.Core/Calculations/ChromaticityBrightnessReference.cs
using IwashiScope.Core.Models;
using static IwashiScope.Core.Calculations.ChromaticityMath;
namespace IwashiScope.Core.Calculations;

/// <summary>White-relative reference scalar, not a prediction of subjective brightness.</summary>
public sealed record ChromaticityBrightnessReference(ConfusionColorMode Mode, double RequestedLuminance)
{
    public double RelativeLuminance => Math.Min(1, RequestedLuminance);
    public bool IsDisplayLimited => RequestedLuminance > 1;
    public static ChromaticityBrightnessReference? Create(Vector3 lms, Vector3 white, ConfusionColorMode mode) =>
        Response(lms, white, mode) is { } value ? new(mode, value) : null;
    public static double? Response(Vector3 lms, Vector3 white, ConfusionColorMode mode)
    {
        if (!Nonnegative(lms) || !Nonnegative(white) || mode == ConfusionColorMode.C) return null;
        var numerator = mode switch { ConfusionColorMode.P => lms.Second, ConfusionColorMode.D => lms.First, _ => Cie2006Chromaticity.XyzF(lms).Second };
        var denominator = mode switch { ConfusionColorMode.P => white.Second, ConfusionColorMode.D => white.First, _ => Cie2006Chromaticity.XyzF(white).Second };
        var value = numerator / denominator;
        return denominator > 0 && double.IsFinite(value) ? value : null;
    }
    public static Vector3 PaletteWhiteLms { get; } = Scale(ChromaticityConfusionBand.LmsFromXyzF(ChromaticityBackgroundPalette.Shared.White.XyzF),
        100 / ChromaticityBackgroundPalette.Shared.White.XyzF.Second);
}

File: IwashiScopePackage/Sources/IwashiScopeFeature/Models/ChromaticityBrightnessReference.swift
import Foundation

/// A white-relative reference signal, not a prediction of subjective brightness.
/// P uses M, D uses L, T uses the existing CIE2015 L/M luminance combination.
struct ChromaticityBrightnessReference: Equatable, Sendable {
    let mode: ConfusionColorMode
    let requestedLuminance: Double
    let relativeLuminance: Double

    var isDisplayLimited: Bool { requestedLuminance > 1 }

    init?(lms: Vector3, whiteLMS: Vector3, mode: ConfusionColorMode) {
        guard mode != .none,
              let value = Self.response(lms, relativeTo: whiteLMS, mode: mode) else { return nil }
        self.mode = mode
        requestedLuminance = value
        relativeLuminance = min(1, value)
    }

    static func response(_ lms: Vector3, relativeTo white: Vector3, mode: ConfusionColorMode) -> Double? {
        guard [lms.first, lms.second, lms.third, white.first, white.second, white.third]
            .allSatisfy({ $0.isFinite && $0 >= 0 }) else { return nil }
        let numerator: Double
        let denominator: Double
        switch mode {
        case .none: return nil
        case .protan: (numerator, denominator) = (lms.second, white.second)
        case .deutan: (numerator, denominator) = (lms.first, white.first)
        case .tritan:
            numerator = CIE2006Chromaticity.xyzF(from: lms).second
            denominator = CIE2006Chromaticity.xyzF(from: white).second
        }
        guard denominator > 0 else { return nil }
        let value = numerator / denominator
        return value.isFinite ? value : nil
    }

    /// The display palette has a fixed full-range D50 white. Keep it distinct from
    /// the measurement's same-range white; normalization bridges these two units.
    static let paletteWhiteLMS: Vector3? = {
        guard let palette = ChromaticityBackgroundPalette.shared else { return nil }
        let value = ChromaticityConfusionBand.lms(fromXYZF: palette.white.xyzF)
        let gain = 100 / palette.white.xyzF.second
        return .init(first: value.first * gain, second: value.second * gain, third: value.third * gain)
    }()

    /// Useful for diagrams of synthetic LMS data with the palette's own white.
    static func forBand(_ band: ChromaticityConfusionBand) -> Self? {
        guard let white = paletteWhiteLMS else { return nil }
        return Self(lms: band.referenceLMS, whiteLMS: white, mode: band.colorMode)
    }
}
